Upgrade express-rate-limit to v8.2.1

alanorth opened the pull request:
# Description
Use the latest version of the express-rate-limit dependency. This version has support for headers conforming to the [RateLimit header fields for HTTP standardization draft](https://github.com/ietf-wg-httpapi/ratelimit-headers) adopted by the IETF. This version also has optional support for express v5 if we migrate to that in the future, improved support for external rate limit stores, and improved support for other similar middleware like [express-slow-down](https://github.com/express-rate-limit/express-slow-down).

See the [long list of changes from version 5.x to 8.x](https://express-rate-limit.mintlify.app/reference/changelog#5-x).

# Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.

List of changes in this PR:
* First, upgrade `express-rate-limit` from `^5.1.3` to `^8.2.1`
* Second, rename the `max` configuration option to `limit`
* Third, enable use of newer standard rate limit headers, see: https://express-rate-limit.mintlify.app/reference/configuration#standardheaders
* Fourth, disable legacy rate limit headers: https://express-rate-limit.mintlify.app/reference/configuration#legacyheaders
* Fifth, allow configuration of the new `ipv6Subnet` setting

**Include guidance for how to test or review your PR.**

* Make sure `ui.rateLimiter.limit` is set to some non-zero value (default 500)
* Start the Angular frontend in production mode with `npm start`
* Make a GET request to any page and look for `RateLimit-*` and other headers in the response (you should _not_ see `X-RateLimit-*`)

# Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

alanorth committed:
Upgrade express-rate-limit to v8.2.1

tdonohue commented on src/config/default-app-config.ts:
```diff
@@ -47,10 +47,10 @@ export class DefaultAppConfig implements AppConfig {
     rateLimiter: {
       windowMs: 1 * 60 * 1000, // 1 minute
-      max: 500, // limit each IP to 500 requests per windowMs
+      limit: 500, // limit each IP to 500 requests per windowMs
```
We may want to take this opportunity to discuss this default limit of 500.  I notice that `express-rate-limit` has a [default of 5](https://express-rate-limit.mintlify.app/reference/configuration#limit), which is 100 times smaller.

It seems odd to me that our limit is so very high.  I wonder if we should consider decreasing this to something like 10 or 20 or even 50 requests per minute.  5 is likely too small, but 500 seems way too large.

mwoodiupui replied:
It's probably that large because we make scads of requests to build many of our "pages".  We are getting better, so maybe we can crank it down a _bit_.

tdonohue replied:
@mwoodiupui : I'm *pretty sure* that `express-rate-limit` is only looking at request to the *User Interface*.  So, I don't believe that REST API requests are counted here (but correct me if I'm wrong).   So, I believe each request here is essentially one SSR-loaded page.  

That's why 500 seems so extreme to me...It's pretty difficult to request 500 SSR-loaded pages in one minute unless you are a very aggressive bot.

tdonohue replied:
Double checked... `express-rate-limit` is only used in [`server.ts`](https://github.com/DSpace/dspace-angular/blob/main/server.ts), and looks to be only enabled for these requests:
* Requests for SSR for all pages
* Requests for `/robots.txt`
* Requests for `/sitemap` paths/files
* Requests for `/signposting` paths/files
* Requests for `/iiif` paths
* Requests for `/app/heath` paths'
* Requests for static resources (images, i18n files, etc) under `/assets/*`

So, it's not quite a 1 SSR page per request (cause of those static resources).  But, 500 still seems large.  The static assets should only be requested once and then cached. After that it's basically one page per request.

Maybe a value of 50 is a more reasonable default?  Most humans shouldn't surpass that unless they are clicking very quickly.

tdonohue reviewed (commented):
Thanks @alanorth !  This looks great so far.  I've added a few inline comments above.  My only other thought here is whether we should add [`ipv6Subnet`](https://express-rate-limit.mintlify.app/reference/configuration#ipv6subnet) as a configurable setting in DSpace?  

It seems like a setting that some sites may want more direct control over in order to better control rate limiting of IP ranges.

alanorth committed:
Allow configuration of express-rate-limit's ipv6Subnet

56 is a moderately aggressive default. It may be increased to if
users are being incorrectly blocked (try 60 or 64), or decreased
if you are seeing evidence of abuse.

See: https://express-rate-limit.mintlify.app/reference/configuration#ipv6subnet

alanorth commented:
Thanks for the comments @mwoodiupui and @tdonohue.

> My only other thought here is whether we should add [ipv6Subnet](https://express-rate-limit.mintlify.app/reference/configuration#ipv6subnet) as a configurable setting in DSpace?

Good idea, I will add it.

mwoodiupui commented on server.ts:
```diff
@@ -179,10 +180,13 @@ export function app() {
    * When it is present, the rateLimiter will be enabled. When it is undefined, the rateLimiter will be disabled.
    */
   if (hasValue((environment.ui as UIServerConfig).rateLimiter)) {
-    const RateLimit = require('express-rate-limit');
```
I guess [dynamic import](https://stackoverflow.com/a/46543835/2916377) won't work here for some reason?

What's the actual cost of the static import?  Maybe it's worth it to comb out another CommonJS leftover?

tdonohue replied:
Good point, @mwoodiupui :  a dynamic import would likely work just as fine here as well.  I was more concerned about the static import, as I believe it will pull in `express-rate-limit` regardless of whether it is being used/enabled by the DSpace site.  

The more we pull in via static imports, the larger our application will be.  I'm not sure of the actual "cost" here, but every place we can use a dynamic import (lazy loading) is *better* for the main application size.

tdonohue replied:
Unresolving this just because we might want to try out the dynamic import approach that @mwoodiupui points out. That's likely even better here, and it should hopefully work seemlessly.

tdonohue replied:
No worries then @alanorth .  I'd say lets leave it as it is then.  It's not as important right now.

tdonohue commented:
@alanorth :  e2e tests run in "production mode", so they won't use `environment.test.ts` at all (so you can remove your changes there).

So, I think you'd instead need to override the default using an environment variable in build.yml... similar to what we do to turn off all caching in all tests here: https://github.com/DSpace/dspace-angular/blob/main/.github/workflows/build.yml#L28-L29

We might be able to get away with setting it to 50 or 100 for e2e tests.  We'll have to see what works better...but I didn't think about the fact that e2e tests are going to be requesting a large number of pages per minute in order to run the tests.

saschaszott commented:
There may be an issue with the rate limit configuration when used in NAT environments.

At our university, all users access external services through a shared public IP address due to NAT. This means that the global threshold (requests per IP address per minute) affects all users collectively, not individually.

If the threshold is set too low, even normal usage by a few users may trigger the rate limit and block legitimate traffic for the entire institution.

While this could be mitigated for authenticated users by applying rate limits per user ID or session, it remains problematic for unauthenticated users, who are all subject to the same global limit.

pnbecker commented:
A limit of 20 might be a problem depending on your local theme. I expect it counts every request, even if the connection stays open with keep a live. If you have a lot of fonts, images and other things and do not use CDNs for privacy reasons, 20 might be low.

alanorth commented:
> _There may be an issue with the rate limit configuration when used in NAT environments._

@saschaszott Yes you are right. The `limit` setting is configurable but we need to find a better default.

> _Would it make sense to use device fingerprinting in combination with the IP address, for example by integrating [fingerprintjs/fingerprintjs](https://github.com/fingerprintjs/fingerprintjs)?_

Maybe, but that particular library has a weird license and even if we found something else, if it runs in JavaScript then only bots that execute JavaScript will use it. I have heard some bots hit item pages (SSR) and scrape the metadata without executing JavaScript or loading any assets.

> _A limit of 20 might be a problem depending on your local theme. I expect it counts every request, even if the connection stays open with keep a live. If you have a lot of fonts, images and other things and do not use CDNs for privacy reasons, 20 might be low._

Yes, fair point. Keep in mind that once the client loads a page once they switch to CSR. Even so, I just loaded the default theme on `main` in a private window and checked the response of the last request and found that I had 2 requests remaining.

<img width="2428" height="1626" alt="2025-08-08T09:58:05,510758206+03:00" src="https://github.com/user-attachments/assets/8d717bc9-f985-4080-81a4-b322ac5c794a" />

So the previous default of 500 is too high, and 20 is too low. Shall we go for 50?

tdonohue commented:
@alanorth : 50 sounds reasonable enough to try out and see if it works better.  As @kshepherd pointed out in yesterday's Developers Meeting, we need to make sure to do testing with browser caching *disabled*.   I believe my initial testing (where I suggested 20) may have been flawed because of my browser caching.  So, we need to try accessing the site as a brand new user, and see what level works best for them.  (I'm about to head on vacation, so I'll leave this to others to do some testing in the meantime)

As a sidenote, you are correct that some crawlers (especially Google Scholar) do not execute Javascript.  I'm not sure whether their crawl will trigger a download of other assets, but we could always ask them as necessary.

pnbecker commented:
I think the approach of a general rate limit is not the best. @alanorth, I think you told me that you were rate limiting access to certain paths like /item or /entities. As long as browser have to make more requests than bots, we will never be able to adjust rate limits correctly. Is there a chance to not rate limit assets but other paths that are accessed by bots without JavaScript? That would also allow us to set rate limits independently from themes.

alanorth commented:
Thanks @tdonohue. I've re-worked this patch series to use the old default of 500 requests, and dropped the override for CI as it is no longer needed.

Shame that this cannot be a useful tool. At this point we should only be upgrading this because it is an aging dependency with its own aging dependencies. Long term I think we should drop it because I've never seen it be useful. Sadly this means that repository admins will have to resort to web application firewalls if they have the skills, or Cloudflare if they have the budget (bad for the open Internet).

One thing I like about the proof of work (PoW) tools like [Anubis](https://github.com/TecharoHQ/anubis/) and [go-away](https://git.gammaspectra.live/git/go-away) is that they have mechanisms to match user agents with known IP ranges. So you can claim you are Googlebot, but unless you come from a known Google network, you get challenged (or blocked). It is non-trivial to deploy though, and comes with other side effects.

nwoodward reviewed (approved):
👍 Thanks @alanorth. I tested several actions with this PR deployed and didn't run into any issues.
